File: algorithms/0020-valid-parentheses/Solution.kt
// Valid Parentheses — stack of opens; each close must match the top open.
// Time O(n), Space O(n).

class Solution {
    fun isValid(s: String): Boolean {
        val pairs = mapOf(')' to '(', ']' to '[', '}' to '{')
        val stack = ArrayDeque<Char>()
        for (ch in s) {
            if (ch !in pairs) {
                stack.push(ch)
                continue
            }
            if (stack.isEmpty() || stack.pop() != pairs[ch]) return false
        }
        return stack.isEmpty()
    }
}

File: algorithms/0020-valid-parentheses/solution.py
# Valid Parentheses — stack of opens; each close must match the top open.
# Time O(n), Space O(n).

class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {")": "(", "]": "[", "}": "{"}
        stack: list[str] = []
        for ch in s:
            if ch not in pairs:
                stack.append(ch)
                continue
            if not stack or stack.pop() != pairs[ch]:
                return False
        return not stack
